README.md
# bitcoinaddress.app
JavaScript Client-Side Bitcoin Wallet Generator

Now Bitcoin addresses and their corresponding private key can be conveniently 
generated in a web browser.

The bitcoinaddress.app project provides an all-in-one HTML document with embedded
JavaScript/Css/Images. The JavaScript is readable not minified and contains no
XMLHttpRequest's (no AJAX). The benefit of this technique is you can load the 
JavaScript locally and trust that the JavaScript did not change after being 
loaded. 

Here is a link to the BitcoinTalk.org forum topic discussing this project:
https://bitcointalk.org/index.php?topic=43496.0


Please send DONATIONS for this project to Bitcoin Address: 
18mNc7PhGa2Ceo7JZezHfUEaBmgVV88ZC1


END USER NOTES:

 1) For Bulk Wallet I recommended using Google Chrome, it's the fastest.

 2) Requires IE9+, Firefox, Chrome or sufficient JavaScript support.

 3) Mobile Safari only works with iPhone4 or newer devices.
    Older devices timeout while executing JavaScript.

 4) DO NOT use Opera Mini it renders JavaScript output server side, therefore
    they might record the private key you generated.

 5) BIP38 most likely will not work on mobile devices due to hardware limitations.


Notice of Copyrights and Licenses:
---------------------------------------
The bitcoinaddress.app project, software and embedded resources are
copyright bitcoinaddress.app.

The bitcoinaddress.app name and logo are not part of the open source
license.

Portions of the all-in-one HTML document contain JavaScript codes that
are the copyrights of others. The individual copyrights are included
throughout the document along with their licenses. Included JavaScript
libraries are separated with HTML script tags.

Summary of JavaScript functions with a redistributable license:

JavaScript function	|	License
-------------------	|	--------------
Array.prototype.map	|	Public Domain
window.Crypto | BSD License
window.SecureRandom	| BSD License
window.EllipticCurve	|	BSD License
window.BigInteger |	BSD License
window.QRCode | MIT License
window.Bitcoin | MIT License

MIT License

Copyright (c) 2011-2016 bitaddress.org (pointbiz)  
Copyright (c) 2021 bitcoinaddress.app (j4em)


Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.txt
bitcoinaddress.app
Open Source JavaScript Client-Side Bitcoin Wallet Generator

The bitcoinaddress.app file contains all the JavaScript/Css/Images embedded
in the HTML document. Included in the file name is the file's version
number and hash.

The bitcoinaddress.app software is available Open Source under the MIT License.
Visit the GitHub Repository for the full copyright and license information:
https://github.com/j4em/bitcoinaddress.app

Bitcoin Forum thread:
https://bitcointalk.org/index.php?topic=43496.0

Donation Address: 18mNc7PhGa2Ceo7JZezHfUEaBmgVV88ZC1

END USER NOTES:
 1) For Bulk Wallet I recommended using Google Chrome, it's the fastest.
 2) Requires IE9+, Firefox, Chrome or sufficient JavaScript support.
 3) Mobile Safari only works with iPhone4 or newer devices.
    Older devices timeout while executing JavaScript.
 4) DO NOT use Opera Mini it renders JavaScript output server side, therefore
    they might record the private key you generated.
 5) BIP38 most likely will not work on mobile devices due to hardware limitations.

Here is a signed list of file names and version history.

2021-04-14: status ACTIVE
bitcoinaddress.app-v4.0.1-SHA256-528933de9433a361f9df058a049108b72dd8258a960b1c0b1af75bc5eee645b0.html
- bitcoinaddress.app
- modifying the code style
- refactor directory structure
- replace 'grunt' with 'inline-stylesheets'

2016-12-23: status ACTIVE
bitaddress.org-v3.3.0-SHA256-dec17c07685e1870960903d8f58090475b25af946fe95a734f88408cef4aa194.html
 - remove support for IE8
 - small translations updates for hu and pt-br

2016-11-27: status ACTIVE
bitaddress.org-v3.2.3-SHA256-c9a0bb3ed50aa75a5ae9c606d81e3fd41a4ff686ad38ad5379e2402f481e79a4.html
 - wallet details: show error when checksum validation fails
 - wallet details: show error when private key outside of curve range

2016-08-21: status ACTIVE
bitaddress.org-v3.2.2-SHA256-f4d047c264a2b71946de319482a9365e56d8d7289dd85a352da3b1448b7647df.html
 - version bump for unix line endings

2016-07-31: status ACTIVE
bitaddress.org-v3.2.1-SHA256-42c3bcb643c451689e5bd1499ed4b516be2da06d2fe3886b0dd15b8fc2525ecd.html
 - BigInteger modInverse should be positive
 - throw if modInverse 0
 - improve BigInteger constructor so that it works if caller forgets 'new'
 - add unit tests for BigInteger
 - thanks to dooglus, jprichardson, dcousens

2016-02-19: status ACTIVE
bitaddress.org-v3.2.0-SHA256-ad4fd171c647772aa76d0ce828731b01ca586596275d43a94008766b758e8736.html
 - switch languages without full page load
 - add BIP38 encryption to Bulk Wallet
 - use compressed addresses on Single/Paper/Bulk Wallet
 - add compressed address option on Brain Wallet

2016-01-17: status ACTIVE
bitaddress.org-v3.1.1-SHA256-8277de0a0c77761caa8f546c9885c36a3134a94823b14e24d364b86abb3a9ab3.html
 - refactor translations into separate files per culture.

2015-11-22: status ACTIVE
bitaddress.org-v3.1.0-SHA256-c3d4d8da8fc6980435a520dff562b7f831b2f6037ec2d4dd6bf76c5321873303.html
 - add BIP38 encryption on Wallet Details tab.

2015-10-25: status ACTIVE
bitaddress.org-v3.0.1-SHA256-24d2d7f047a9aa217bf69f3ef344c972c151b1e3f6a8aa86ceb9a3be62884bc0.html
 - fix for session log not keeping track of keys from "Wallet Details" tab before entropy is collected.

2015-10-25: status ACTIVE
bitaddress.org-v3.0.0-SHA256-4781574ca09c07f65d1966619f37a762aac6decd8732cacc85b2f2f972f82751.html
 - add session log icon that shows all the key pairs generated during the current session.

2015-08-16: status ACTIVE
bitaddress.org-v2.9.11-SHA256-40376eddc790a63d9afcfb72c0a45002827da965f3bfe6ba8c330e697bf188b2.html
 - add status icons for checking the URI protocol used, support for window.crypto.getRandomValues
   and run the synchronous unit tests after entropy collection.

2015-07-18: status ACTIVE
bitaddress.org-v2.9.10-SHA256-445e44cfd04c8f1ea8f732c3ae7277b0166fdb3e2109251c54e4b367983fe04d.html
 - add Portuguese. Thanks rhcastilhos.
 - minor french updates.

2015-07-05: status ACTIVE
bitaddress.org-v2.9.9-SHA256-90ddaf250f6302acb53945128e38225208af5a2fa7cfdf51519213e8b144a76d.html
 - improve tab usability. You can now get to the Brain Wallet and Wallet Details tabs before
   completing the entropy collection.

2015-01-08: status ACTIVE
bitaddress.org-v2.9.8-SHA256-2c5d16dbcde600147162172090d940fd9646981b7d751d9bddfc5ef383f89308.html
 - fix French translations. Escape quotes.

2015-01-07: status OFFLINE
bitaddress.org-v2.9.7-SHA256-1b0f71dfc2e064426328c15c4dbd1f467cb26afe0e84841347ad11d8ca668f70.html
 - fix translations

2015-01-05: status ACTIVE
bitaddress.org-v2.9.6-SHA256-34728a9cec417cf8060423c77e8793e4aa133cf3d66a6b8073b4290cf1f4695c.html
 - Japanese translations for Split Wallet. Thanks dabura667.
 - remove promise to show MINI key on details tab. MINI key
   will only be shown when it is provided since it cannot
   be derived from other key formats.
 - fix README
 - Russian translations. Thanks e5faf2.
 - Simplified Chinese translations. Thanks kwl01skz.
 - add direct link to zip on github
 - add this CHANGELOG to repository and add detached sigs in repository.
   add link to sig of HTML.
 - hash with SHA256 instead of SHA1. SHA1 hash still provided in package.json.

2014-04-23: status ACTIVE
bitaddress.org-v2.9.3-SHA1-7d47ab312789b7b3c1792e4abdb8f2d95b726d64.html
 - increased the HTML height to allow for greater range of mouse
   seed values on large monitors. Thanks danbartram.
 - Japanese translations. Thanks dabura667.

2014-04-21: status ACTIVE
bitaddress.org-v2.9.1-SHA1-67b1facd70890aa9544597e97122c7a1d4fdc821.html
 - Hungarian translation. Thanks bitcoin333.
 - Auto detect culture and translate. Thanks onovy.

2014-04-15: status ACTIVE
bitaddress.org-v2.9.0-SHA1-6e9ae5c64d510b53fa39e36a3017d5342b838984.html
 - Split Wallet: Shamir's Secret Sharing for a Bitcoin private key.
   Thanks to Jeff Weiss.

2014-01-19: status ACTIVE

src/secrets.js
// secrets.js - by Alexander Stetsyuk - released under MIT License
(function(exports, global){
var defaults = {
    bits: 8, // default number of bits
    radix: 16, // work with HEX by default
    minBits: 3,
    maxBits: 20, // this permits 1,048,575 shares, though going this high is NOT recommended in JS!

    bytesPerChar: 2,
    maxBytesPerChar: 6, // Math.pow(256,7) > Math.pow(2,53)

    // Primitive polynomials (in decimal form) for Galois Fields GF(2^n), for 2 <= n <= 30
    // The index of each term in the array corresponds to the n for that polynomial
    // i.e. to get the polynomial for n=16, use primitivePolynomials[16]
    primitivePolynomials: [null,null,1,3,3,5,3,3,29,17,9,5,83,27,43,3,45,9,39,39,9,5,3,33,27,9,71,39,9,5,83],

    // warning for insecure PRNG
    warning: 'WARNING:\nA secure random number generator was not found.\nUsing Math.random(), which is NOT cryptographically strong!'
};

// Protected settings object
var config = {};

/** @expose **/
exports.getConfig = function(){
    return {
        'bits': config.bits,
        'unsafePRNG': config.unsafePRNG
    };
};

function init(bits){
    if(bits && (typeof bits !== 'number' || bits%1 !== 0 || bits<defaults.minBits || bits>defaults.maxBits)){
        throw new Error('Number of bits must be an integer between ' + defaults.minBits + ' and ' + defaults.maxBits + ', inclusive.')
    }

    config.radix = defaults.radix;
    config.bits = bits || defaults.bits;
    config.size = Math.pow(2, config.bits);
    config.max = config.size - 1;

    // Construct the exp and log tables for multiplication.
    var logs = [], exps = [], x = 1, primitive = defaults.primitivePolynomials[config.bits];
    for(var i=0; i<config.size; i++){
        exps[i] = x;
        logs[x] = i;
        x <<= 1;
        if(x >= config.size){
            x ^= primitive;
            x &= config.max;
        }
    }

    config.logs = logs;
    config.exps = exps;
};

/** @expose **/
exports.init = init;

function isInited(){
    if(!config.bits || !config.size || !config.max  || !config.logs || !config.exps || config.logs.length !== config.size || config.exps.length !== config.size){
        return false;
    }
    return true;
};

// Returns a pseudo-random number generator of the form function(bits){}
// which should output a random string of 1's and 0's of length `bits`
function getRNG(){
    var randomBits, crypto;

    function construct(bits, arr, radix, size){
        var str = '',
            i = 0,
            len = arr.length-1;
        while( i<len || (str.length < bits) ){
            str += padLeft(parseInt(arr[i], radix).toString(2), size);
            i++;
        }
        str = str.substr(-bits);
        if( (str.match(/0/g)||[]).length === str.length){ // all zeros?
            return null;
        }else{
            return str;
        }
    }

    // node.js crypto.randomBytes()
    if(typeof require === 'function' && (crypto=require('crypto')) && (randomBits=crypto['randomBytes'])){
        return function(bits){
            var bytes = Math.ceil(bits/8),
                str = null;

            while( str === null ){
                str = construct(bits, randomBits(bytes).toString('hex'), 16, 4);
            }
            return str;
        }
    }

    // browsers with window.crypto.getRandomValues()
    if(global['crypto'] && typeof global['crypto']['getRandomValues'] === 'function' && typeof global['Uint32Array'] === 'function'){
        crypto = global['crypto'];
        return function(bits){
            var elems = Math.ceil(bits/32),
                str = null,
                arr = new global['Uint32Array'](elems);

            while( str === null ){
                crypto['getRandomValues'](arr);
                str = construct(bits, arr, 10, 32);
            }

            return str;
        }
    }

    // A totally insecure RNG!!! (except in Safari)
    // Will produce a warning every time it is called.
    config.unsafePRNG = true;
    warn();

    var bitsPerNum = 32;
    var max = Math.pow(2,bitsPerNum)-1;
    return function(bits){
        var elems = Math.ceil(bits/bitsPerNum);
        var arr = [], str=null;
        while(str===null){
            for(var i=0; i<elems; i++){
                arr[i] = Math.floor(Math.random() * max + 1);
            }
            str = construct(bits, arr, 10, bitsPerNum);
        }
        return str;
    };
};

// Warn about using insecure rng.
// Called when Math.random() is being used.
function warn(){
    global['console']['warn'](defaults.warning);
    if(typeof global['alert'] === 'function' && config.alert){
        global['alert'](defaults.warning);
    }
}

// Set the PRNG to use. If no RNG function is supplied, pick a default using getRNG()
/** @expose **/
exports.setRNG = function(rng, alert){
    if(!isInited()){
        this.init();
    }
    config.unsafePRNG=false;
    rng = rng || getRNG();

    // test the RNG (5 times)
    if(typeof rng !== 'function' || typeof rng(config.bits) !== 'string' || !parseInt(rng(config.bits),2) || rng(config.bits).length > config.bits || rng(config.bits).length < config.bits){
        throw new Error("Random number generator is invalid. Supply an RNG of the form function(bits){} that returns a string containing 'bits' number of random 1's and 0's.")
    }else{
        config.rng = rng;
    }
    config.alert = !!alert;

    return !!config.unsafePRNG;
};

function isSetRNG(){
    return typeof config.rng === 'function';
};

// Generates a random bits-length number string using the PRNG
/** @expose **/
exports.random = function(bits){
    if(!isSetRNG()){
        this.setRNG();
    }

    if(typeof bits !== 'number' || bits%1 !== 0 || bits < 2){
        throw new Error('Number of bits must be an integer greater than 1.')
    }

    if(config.unsafePRNG){
        warn();
    }
    return bin2hex(config.rng(bits));
}

// Divides a `secret` number String str expressed in radix `inputRadix` (optional, default 16)
// into `numShares` shares, each expressed in radix `outputRadix` (optional, default to `inputRadix`),
// requiring `threshold` number of shares to reconstruct the secret.
// Optionally, zero-pads the secret to a length that is a multiple of padLength before sharing.
/** @expose **/
exports.share = function(secret, numShares, threshold, padLength, withoutPrefix){
    if(!isInited()){
        this.init();
    }
    if(!isSetRNG()){
        this.setRNG();
    }

    padLength =  padLength || 0;

    if(typeof secret !== 'string'){
        throw new Error('Secret must be a string.');
    }
    if(typeof numShares !== 'number' || numShares%1 !== 0 || numShares < 2){
        throw new Error('Number of shares must be an integer between 2 and 2^bits-1 (' + config.max + '), inclusive.')
    }
    if(numShares > config.max){
        var neededBits = Math.ceil(Math.log(numShares +1)/Math.LN2);
        throw new Error('Number of shares must be an integer between 2 and 2^bits-1 (' + config.max + '), inclusive. To create ' + numShares + ' shares, use at least ' + neededBits + ' bits.')
    }
    if(typeof threshold !== 'number' || threshold%1 !== 0 || threshold < 2){
        throw new Error('Threshold number of shares must be an integer between 2 and 2^bits-1 (' + config.max + '), inclusive.');
    }
    if(threshold > config.max){
        var neededBits = Math.ceil(Math.log(threshold +1)/Math.LN2);
        throw new Error('Threshold number of shares must be an integer between 2 and 2^bits-1 (' + config.max + '), inclusive.  To use a threshold of ' + threshold + ', use at least ' + neededBits + ' bits.');
    }
    if(typeof padLength !== 'number' || padLength%1 !== 0 ){
        throw new Error('Zero-pad length must be an integer greater than 1.');
    }

    if(config.unsafePRNG){
        warn();
    }

    secret = '1' + hex2bin(secret); // append a 1 so that we can preserve the correct number of leading zeros in our secret
    secret = split(secret, padLength);
    var x = new Array(numShares), y = new Array(numShares);
    for(var i=0, len = secret.length; i<len; i++){
        var subShares = this._getShares(secret[i], numShares, threshold);
        for(var j=0; j<numShares; j++){
            x[j] = x[j] || subShares[j].x.toString(config.radix);
            y[j] = padLeft(subShares[j].y.toString(2)) + (y[j] ? y[j] : '');
        }
    }
    var padding = config.max.toString(config.radix).length;
    if(withoutPrefix){
        for(var i=0; i<numShares; i++){
            x[i] = bin2hex(y[i]);
        }
    }else{
        for(var i=0; i<numShares; i++){
            x[i] = config.bits.toString(36).toUpperCase() + padLeft(x[i],padding) + bin2hex(y[i]);
        }
    }

    return x;
};

// This is the basic polynomial generation and evaluation function
// for a `config.bits`-length secret (NOT an arbitrary length)
// Note: no error-checking at this stage! If `secrets` is NOT
// a NUMBER less than 2^bits-1, the output will be incorrect!
/** @expose **/
exports._getShares = function(secret, numShares, threshold){
    var shares = [];
    var coeffs = [secret];

    for(var i=1; i<threshold; i++){
        coeffs[i] = parseInt(config.rng(config.bits),2);
    }
    for(var i=1, len = numShares+1; i<len; i++){
        shares[i-1] = {
            x: i,
            y: horner(i, coeffs)
        }
    }
    return shares;
};

// Polynomial evaluation at `x` using Horner's Method
// TODO: this can possibly be sped up using other methods
// NOTE: fx=fx * x + coeff[i] ->  exp(log(fx) + log(x)) + coeff[i],
//       so if fx===0, just set fx to coeff[i] because
//       using the exp/log form will result in incorrect value
function horner(x, coeffs){
    var logx = config.logs[x];
    var fx = 0;
    for(var i=coeffs.length-1; i>=0; i--){
        if(fx === 0){
            fx = coeffs[i];
            continue;
        }
        fx = config.exps[ (logx + config.logs[fx]) % config.max ] ^ coeffs[i];
    }
    return fx;
};

function inArray(arr,val){
    for(var i = 0,len=arr.length; i < len; i++) {
        if(arr[i] === val){
            return true;
         }
     }
    return false;
};

function processShare(share){

    var bits = parseInt(share[0], 36);
    if(bits && (typeof bits !== 'number' || bits%1 !== 0 || bits<defaults.minBits || bits>defaults.maxBits)){
        throw new Error('Number of bits must be an integer between ' + defaults.minBits + ' and ' + defaults.maxBits + ', inclusive.')
    }

    var max = Math.pow(2, bits) - 1;
    var idLength = max.toString(config.radix).length;

    var id = parseInt(share.substr(1, idLength), config.radix);
    if(typeof id !== 'number' || id%1 !== 0 || id<1 || id>max){
        throw new Error('Share id must be an integer between 1 and ' + config.max + ', inclusive.');
    }
    share = share.substr(idLength + 1);
    if(!share.length){
        throw new Error('Invalid share: zero-length share.')
    }
    return {
        'bits': bits,
        'id': id,
        'value': share
    };
};

/** @expose **/
exports._processShare = processShare;

// Protected method that evaluates the Lagrange interpolation
// polynomial at x=`at` for individual config.bits-length
// segments of each share in the `shares` Array.
// Each share is expressed in base `inputRadix`. The output
// is expressed in base `outputRadix'
function combine(at, shares){
    var setBits, share, x = [], y = [], result = '', idx;

    for(var i=0, len = shares.length; i<len; i++){
        share = processShare(shares[i]);
        if(typeof setBits === 'undefined'){
            setBits = share['bits'];
        }else if(share['bits'] !== setBits){
            throw new Error('Mismatched shares: Different bit settings.')
        }

        if(config.bits !== setBits){
            init(setBits);
        }

        if(inArray(x, share['id'])){ // repeated x value?
            continue;
        }

        idx = x.push(share['id']) - 1;
        share = split(hex2bin(share['value']));
        for(var j=0, len2 = share.length; j<len2; j++){
            y[j] = y[j] || [];
            y[j][idx] = share[j];
        }
    }

    for(var i=0, len=y.length; i<len; i++){
        result = padLeft(lagrange(at, x, y[i]).toString(2)) + result;
    }

    if(at===0){// reconstructing the secret
        var idx = result.indexOf('1'); //find the first 1
        return bin2hex(result.slice(idx+1));
    }else{// generating a new share
        return bin2hex(result);
    }
};

// Combine `shares` Array into the original secret
/** @expose **/
exports.combine = function(shares){
    return combine(0, shares);
};

// Generate a new share with id `id` (a number between 1 and 2^bits-1)
// `id` can be a Number or a String in the default radix (16)
/** @expose **/
exports.newShare = function(id, shares){
    if(typeof id === 'string'){
        id = parseInt(id, config.radix);
    }

    var share = processShare(shares[0]);
    var max = Math.pow(2, share['bits']) - 1;

    if(typeof id !== 'number' || id%1 !== 0 || id<1 || id>max){
        throw new Error('Share id must be an integer between 1 and ' + config.max + ', inclusive.');
    }

    var padding = max.toString(config.radix).length;
    return config.bits.toString(36).toUpperCase() + padLeft(id.toString(config.radix), padding) + combine(id, shares);
};

// Evaluate the Lagrange interpolation polynomial at x = `at`
// using x and y Arrays that are of the same length, with
// corresponding elements constituting points on the polynomial.
function lagrange(at, x, y){
    var sum = 0,
        product,
        i, j;

    for(var i=0, len = x.length; i<len; i++){
        if(!y[i]){
            continue;
        }

        product = config.logs[y[i]];
        for(var j=0; j<len; j++){
            if(i === j){ continue; }
            if(at === x[j]){ // happens when computing a share that is in the list of shares used to compute it
                product = -1; // fix for a zero product term, after which the sum should be sum^0 = sum, not sum^1
                break;
            }
            product = ( product + config.logs[at ^ x[j]] - config.logs[x[i] ^ x[j]] + config.max/* to make sure it's not negative */ ) % config.max;
        }

        sum = product === -1 ? sum : sum ^ config.exps[product]; // though exps[-1]= undefined and undefined ^ anything = anything in chrome, this behavior may not hold everywhere, so do the check
    }
    return sum;
};

/** @expose **/
exports._lagrange = lagrange;

// Splits a number string `bits`-length segments, after first
// optionally zero-padding it to a length that is a multiple of `padLength.
// Returns array of integers (each less than 2^bits-1), with each element
// representing a `bits`-length segment of the input string from right to left,
// i.e. parts[0] represents the right-most `bits`-length segment of the input string.
function split(str, padLength){
    if(padLength){
        str = padLeft(str, padLength)
    }
    var parts = [];
    for(var i=str.length; i>config.bits; i-=config.bits){
        parts.push(parseInt(str.slice(i-config.bits, i), 2));
    }
    parts.push(parseInt(str.slice(0, i), 2));
    return parts;
};

// Pads a string `str` with zeros on the left so that its length is a multiple of `bits`
function padLeft(str, bits){
    bits = bits || config.bits
    var missing = str.length % bits;
    return (missing ? new Array(bits - missing + 1).join('0') : '') + str;
};

function hex2bin(str){
    var bin = '', num;
    for(var i=str.length - 1; i>=0; i--){
        num = parseInt(str[i], 16)
        if(isNaN(num)){
            throw new Error('Invalid hex character.')
        }
        bin = padLeft(num.toString(2), 4) + bin;
    }
    return bin;
}

function bin2hex(str){
    var hex = '', num;
    str = padLeft(str, 4);
    for(var i=str.length; i>=4; i-=4){
        num = parseInt(str.slice(i-4, i), 2);
        if(isNaN(num)){
            throw new Error('Invalid binary character.')
        }
        hex = num.toString(16) + hex;
    }
    return hex;
}

// Converts a given UTF16 character string to the HEX representation.
// Each character of the input string is represented by
// `bytesPerChar` bytes in the output string.
/** @expose **/
exports.str2hex = function(str, bytesPerChar){
    if(typeof str !== 'string'){
        throw new Error('Input must be a character string.');
    }
    bytesPerChar = bytesPerChar || defaults.bytesPerChar;

    if(typeof bytesPerChar !== 'number' || bytesPerChar%1 !== 0 || bytesPerChar<1 || bytesPerChar > defaults.maxBytesPerChar){
        throw new Error('Bytes per character must be an integer between 1 and ' + defaults.maxBytesPerChar + ', inclusive.')
    }

    var hexChars = 2*bytesPerChar;
    var max = Math.pow(16, hexChars) - 1;
    var out = '', num;
    for(var i=0, len=str.length; i<len; i++){
        num = str[i].charCodeAt();
        if(isNaN(num)){
            throw new Error('Invalid character: ' + str[i]);
        }else if(num > max){
            var neededBytes = Math.ceil(Math.log(num+1)/Math.log(256));
            throw new Error('Invalid character code (' + num +'). Maximum allowable is 256^bytes-1 (' + max + '). To convert this character, use at least ' + neededBytes + ' bytes.')
        }else{
            out = padLeft(num.toString(16), hexChars) + out;
        }
    }
    return out;
};

// Converts a given HEX number string to a UTF16 character string.
/** @expose **/
exports.hex2str = function(str, bytesPerChar){
    if(typeof str !== 'string'){
        throw new Error('Input must be a hexadecimal string.');
    }
    bytesPerChar = bytesPerChar || defaults.bytesPerChar;

    if(typeof bytesPerChar !== 'number' || bytesPerChar%1 !== 0 || bytesPerChar<1 || bytesPerChar > defaults.maxBytesPerChar){
        throw new Error('Bytes per character must be an integer between 1 and ' + defaults.maxBytesPerChar + ', inclusive.')
    }

    var hexChars = 2*bytesPerChar;
    var out = '';
    str = padLeft(str, hexChars);
    for(var i=0, len = str.length; i<len; i+=hexChars){
        out = String.fromCharCode(parseInt(str.slice(i, i+hexChars),16)) + out;
    }
    return out;
};

// by default, initialize without an RNG
exports.init();
})(typeof module !== 'undefined' && module['exports'] ? module['exports'] : (window['secrets'] = {}), typeof GLOBAL !== 'undefined' ? GLOBAL : window );